### PR TITLE
Expose Hash, Sign, VerifySignature and VerifyHashSignature

### PR DESCRIPTION
**Pull request by jvehent**
This patch refactors some of the internal logic to expose public functions that can be used outside of a MAR file.

### CODE REVIEW

File: sign.go [hunk 1]
```diff
@@ -64,44 +66,16 @@ func (file *File) FinalizeSignatures() error {
 	if err != nil {
 		return err
 	}
+	if len(file.Signatures) == 0 {
+		return fmt.Errorf("there are no signature to finalize")
```
Comment (g-k) [inline sign.go]: nit: s/signature/signatures/

File: sign.go [hunk 2]
```diff
@@ -64,44 +66,16 @@ func (file *File) FinalizeSignatures() error {
...
-		case SigAlgRsaPkcs1Sha384, SigAlgEcdsaP384Sha384:
-			md = sha512.New384()
-		}
-		md.Write(signableBlock)
-
-		// call the signer interface of the private key to sign the hash
-		sigData, err := file.Signatures[i].privateKey.(crypto.Signer).Sign(
-			rand.Reader, md.Sum(nil), h)
+		hashed, hashAlg, err := Hash(signableBlock, file.Signatures[i].AlgorithmID)
+		sigData, err := Sign(file.Signatures[i].privateKey, rand.Reader, hashed, hashAlg)
```
Comment (g-k) [inline sign.go]: :+1:

File: sign.go [hunk 3]
```diff
@@ -112,6 +86,46 @@ func (file *File) MarshalForSignature() ([]byte, error) {
 	return file.Marshal()
 }
 
+// Hash takes an input and a signature algorithm and returns its hashed value
+func Hash(input []byte, sigalg uint32) (output []byte, h crypto.Hash, err error) {
```
Comment (g-k) [inline sign.go]: might want a type alias for the sig algs

File: sign.go [hunk 4]
```diff
@@ -112,6 +86,46 @@ func (file *File) MarshalForSignature() ([]byte, error) {
...
+		return nil, h, fmt.Errorf("unsupported signature algorithm")
+	}
+	md.Write(input)
+	return md.Sum(nil), h, nil
+}
+
+// Sign signs digest with the private key, possibly using entropy from rand
+func Sign(key crypto.PrivateKey, rand io.Reader, digest []byte, h crypto.Hash) (sigData []byte, err error) {
+	// call the signer interface of the private key to sign the hash
+	sigData, err = key.(crypto.Signer).Sign(rand, digest, h)
```
Comment (g-k) [inline sign.go]: can the cast to crypto.Signer fail and if so do we need to handle that?

File: verify.go [hunk 5]
```diff
@@ -4,16 +4,44 @@ import (
...
+func VerifySignature(input []byte, signature []byte, sigalg uint32, key crypto.PublicKey) error {
+	digest, hashAlg, err := Hash(input, sigalg)
+	if err != nil {
+		return err
+	}
+	return VerifyHashSignature(signature, digest, hashAlg, key)
+}
+
+// VerifyHashSignature takes a signature, the digest of a signed MAR block, a hash algorithm and a public
+// key and return nil if a valid signature is found, or an error if it isn't
```
Comment (g-k) [inline verify.go]: nit: s/return/returns/

File: verify.go [hunk 6]
```diff
@@ -4,16 +4,44 @@ import (
...
+func VerifyHashSignature(signature []byte, digest []byte, hashAlg crypto.Hash, key crypto.PublicKey) error {
```
Comment (g-k) [inline verify.go]: :+1: for splitting this out of VerifySignature

File: sign.go (hunk 3 above)
Comment (jvehent) [reply sign.go]: does that buy us anything? we'll need to type assert to uint32 for marshalling

File: sign.go (hunk 4 above)
Comment (jvehent) [reply sign.go]: I added an assertion that will return an error if Signer isn't implemented

File: sign.go (hunk 3 above)
Comment (g-k) [reply sign.go]: it's more explicit

**Pull request merged**